pilot/pkg/ambient/controller/waypoint_controller.go

```diff
@@ -127,106 +140,89 @@ func (rc *WaypointProxyController) Reconcile(name types.NamespacedName) error {
+		List(klabels.Everything())
 	for _, sa := range serviceAccounts {
 		if gatewaySA != "" && sa.Name != gatewaySA {
-			log.Debugf("skip service account %v, doesn't match gateway %v", sa.Name, gatewaySA)
+			istiolog.Debugf("skip service account %v, doesn't match gateway %v", sa.Name, gatewaySA)
 			continue
 		}
 		wantProxies.Insert(sa.Name)
 	}
 
-	add, remove := wantProxies.Diff(haveProxies)
-
-	if len(remove)+len(add) == 0 {
-		log.Debugf("reconcile: remove %d, add %d. Have %d", len(remove), len(add), len(haveProxies))
-	} else {
-		log.Infof("reconcile: remove %d, add %d. Have %d", len(remove), len(add), len(haveProxies))
-	}
-	for _, k := range remove {
-		log.Infof("removing waypoint proxy %q", k+"-waypoint-proxy")
-		if err := rc.client.Kube().AppsV1().Deployments(gw.Namespace).Delete(context.Background(), k+"-waypoint-proxy", metav1.DeleteOptions{}); err != nil {
-			return fmt.Errorf("pod remove: %v", err)
-		}
-
-		msg := fmt.Sprintf("Removed waypoint proxy from %q namespace", gw.Namespace)
-		if gatewaySA != "" {
-			msg += fmt.Sprintf(" for %q service account", gatewaySA)
-		}
-		err := rc.UpdateStatus(gw, map[string]*istiogw.Condition{
-			string(v1alpha2.GatewayConditionAccepted): {
-				Reason:  string(v1alpha2.GatewayReasonAccepted),
-				Message: msg,
-			},
-		})
-		if err != nil {
-			log.Errorf("unable to update Gateway status %v on delete: %v", gw.Name, err)
-		}
-	}
-	for _, k := range add {
-		log.Infof("adding waypoint proxy %v", k+"-waypoint-proxy")
+	for _, sa := range wantProxies.UnsortedList() {
 		input := MergedInput{
 			Namespace:      gw.Namespace,
 			GatewayName:    gw.Name,
 			UID:            string(gw.UID),
-			ServiceAccount: k,
+			ServiceAccount: sa,
 			Cluster:        rc.cluster.String(),
 			ProxyConfig:    rc.injectConfig().MeshConfig.GetDefaultConfig(),
 		}
-		proxyDeploy, err := rc.RenderDeploymentMerged(input)
+		proxyDeploy, err := rc.RenderDeploymentApply(input)
 		if err != nil {
 			return err
 		}
-
-		if _, err := rc.client.Kube().AppsV1().Deployments(name.Namespace).Create(context.Background(), proxyDeploy, metav1.CreateOptions{}); err != nil {
-			return fmt.Errorf("pod create: %v", err)
-		}
-
-		msg := fmt.Sprintf("Deployed waypoint proxy to %q namespace", gw.Namespace)
-		if gatewaySA != "" {
-			msg += fmt.Sprintf(" for %q service account", gatewaySA)
-		}
-		err = rc.UpdateStatus(gw, map[string]*istiogw.Condition{
-			string(v1alpha2.GatewayConditionReady): {
-				Reason:  string(v1alpha2.GatewayReasonReady),
-				Message: msg,
-			},
-			string(v1alpha2.GatewayConditionAccepted): {
-				Reason:  string(v1alpha2.GatewayReasonAccepted),
-				Message: msg,
-			},
-		})
+		// TODO: should support HPA, PDB, maybe others...
+		istiolog.Infof("Updating waypoint proxy %q", sa+"-waypoint-proxy")
+		_, err = rc.client.Kube().
+			AppsV1().
+			Deployments(name.Namespace).
+			Apply(context.Background(), proxyDeploy, metav1.ApplyOptions{
```

Comment:
What is the benefit of RenderDeploymentApply not just getting yaml/json and using `types.ApplyPatchType`? That is what we do in deploymentcontroller.

Although I will say my reason for not using it (https://github.com/istio/istio/blob/c874e8472f28322152c8d76b938585ab35cab345/pilot/pkg/config/kube/gateway/deploymentcontroller.go#L68) may be a bit short sighted

Reply:
That would produce an identical API call, but with a much more complicated use of client-go.  Since apply is supported in this library, it seems to make sense to use it.  I can change this to hand craft an apply using the dynamic client if you have a strong preference...